firebase-common/src/commonMain/kotlin/dev/gitlive/firebase/Polymorphic.kt

```diff
@@ -0,0 +1,59 @@
+package dev.gitlive.firebase
+
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.SerializationStrategy
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.findPolymorphicSerializer
+import kotlinx.serialization.internal.AbstractPolymorphicSerializer
+
+/*
+ * This code was inspired on polymorphic json serialization of kotlinx.serialization.
+ * See https://github.com/Kotlin/kotlinx.serialization/blob/master/formats/json/commonMain/src/kotlinx/serialization/json/internal/Polymorphic.kt
+ */
+@Suppress("UNCHECKED_CAST")
+internal fun <T> FirebaseEncoder.encodePolymorphically(
+    serializer: SerializationStrategy<T>,
+    value: T,
+    settings: EncodeSettings,
+    ifPolymorphic: (String) -> Unit
+) {
+    if (serializer !is AbstractPolymorphicSerializer<*> || settings.polymorphicStructure == EncodeDecodeSettings.PolymorphicStructure.LIST) {
```

Comment:
please add comments on why the logic is exactly like this. why `LIST` is treated this way